hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/topic/SubscriptionTopicPayloadBuilder.java

```diff
@@ -91,6 +125,93 @@ public IBaseBundle buildPayload(
 		return retval;
 	}
 
+	private Set<IBaseResource> getNotificationShapeResources(List<IBaseResource> theResources, String theTopicUrl) {
+
+		Set<IBaseResource> resultResources = new HashSet<>();
+
+		if (theResources.isEmpty() || mySubscriptionTopicRegistry == null) {
+			return Set.of();
+		}
+
+		// Get the topic from the registry
+		Optional<SubscriptionTopic> oTopic = mySubscriptionTopicRegistry.findSubscriptionTopicByUrl(theTopicUrl);
+		if (oTopic.isEmpty()) {
+			ourLog.debug("No subscription topic found for URL: {}", theTopicUrl);
+			return Set.of();
+		}
+		SubscriptionTopic topic = oTopic.get();
+
+		// Process each notification shape
+		for (SubscriptionTopic.SubscriptionTopicNotificationShapeComponent shape : topic.getNotificationShape()) {
+			String resourceType = shape.getResource();
+
+			// If the resourceType doesn't match any of our resources, skip it
+			List<IBaseResource> resourcesOfThisType = theResources.stream()
+					.filter(r -> myFhirContext.getResourceType(r).equals(resourceType))
+					.collect(Collectors.toList());
+
+			if (resourcesOfThisType.isEmpty()) {
+				continue;
+			}
+
+			List<StringType> include = shape.getInclude();
+			List<StringType> revInclude = shape.getRevInclude();
+
+			Set<IBaseResource> includedResources =
+					getIncludedResources(resourceType, resourcesOfThisType, include, revInclude);
+			resultResources.addAll(includedResources);
+		}
+
+		return resultResources;
+	}
+
+	private Set<IBaseResource> getIncludedResources(
+			String theResourceType,
+			List<IBaseResource> theResourcesOfThisType,
+			List<StringType> theIncludes,
+			List<StringType> theRevIncludes) {
+		Set<IBaseResource> resultResources = new HashSet<>();
+		IFhirResourceDao<?> dao = myDaoRegistry.getResourceDao(theResourceType);
+
+		for (IBaseResource resource : theResourcesOfThisType) {
+			StringBuilder query = new StringBuilder(theResourceType + "?" + PARAM_ID + "="
+					+ resource.getIdElement().getIdPart());
+			for (StringType include : theIncludes) {
+				query.append("&" + PARAM_INCLUDE + "=").append(include.getValue());
```

Comment:
Avoid any possibllity of injection attacks (I can't actually think of how you'd do one here, but people are clever..):

```suggestion
				query.append("&" + PARAM_INCLUDE + "=").append(ParameterUtil.escapeAndUrlEncode(include.getValue()));
```